packages/template-retail-react-app/app/components/icons/index.jsx

```diff
@@ -95,8 +100,9 @@ export const icon = (name, passProps, localizationAttributes) => {
         .replace(/-/g, '')
     const component = forwardRef((props, ref) => {
         const theme = useTheme()
-        const intl = useIntl()
-        if (localizationAttributes) {
+        if (localizationAttributes && props?.intl) {
+            const {intl, ...otherProps} = props
+            props = otherProps
```

Comment:
We don't necessarily need `props.intl`! This will only throw if `<IntlProvider>` is not present _and_ we need it. (Whereas in the current state, we throw if it's missing even if we _don't_ need it.)

```suggestion
        // NOTE: We want to avoid `useIntl` here because that throws when <IntlProvider> is not in
        // the component ancestry, but we only enforce `intl` if we have `localizationAttributes`.
        const intl = useContext(IntlContext)
        if (localizationAttributes) {
            // Use a similar message to the one thrown by `useIntl`.
            if (!intl) {
                throw new Error(
                    'To localize messages, <IntlProvider> needs to exist in the component ancestry.'
                )
            }
```

This is sufficient to address the specific case of the Error component, because the Error component does not use translations. However, if we want provide `props.intl` as a fallback, we can do this:

```suggestion

        // NOTE: We want to avoid `useIntl` here because that throws when <IntlProvider> is not in
        // the component ancestry, but we only enforce `intl` if we have `localizationAttributes`.
        let intl = useContext(IntlContext)
        if (localizationAttributes) {
            // Allow `props.intl` to take precedence over the intl we found
            if (props?.intl) intl = props.intl
            if (!intl) {
                throw new Error(
                    'To localize messages, you must either have <IntlProvider> in the component ancestry or provide `intl` as a prop'
                )
            }
```

I'm not convinced that supporting `props.intl` is necessary, though. If the parent prop has access to the `<IntlProvider>`, then wouldn't the child (this icon) have access as well? In other words, we definitely need to solve "both have access" and "neither have access"; my first suggestion does this without needing an extra prop. Do we need to also solve "parent has access, child does not"? We only need `props.intl` if the answer to that is yes.

---

For reference, here's the [implementation of `useIntl`](https://github.com/formatjs/formatjs/blob/react-intl%405.25.1/packages/react-intl/src/components/useIntl.ts), so you can see that `useContext(IntlProvider)` is the same thing, just without throwing on `null`.